01_homework/hw2_task4.cpp

```diff
@@ -0,0 +1,183 @@
+    std::array<int, population> ownership;
+    std::array<int, population> starvation;
+
+    void print(bool lock, int id);
+    bool check_neighbour_stavation(int id);
+};
+
+#ifdef LOG
+void log_print(int id, std::string_view text)
+{
+   std::lock_guard<std::mutex> cout_lock(print_mutex);
+   std::cout << static_cast<char>(id+65) << text << std::endl;
+}
+#endif
+
+std::string print_fork(int value) {
+    if (value < 0) return "|  ";
+    else if (value == 0) return " - ";
+    else return "  |";
+}
+
+void Table::print(bool lock, int id) {
+    std::lock_guard<std::mutex> print_lock(print_mutex);
+
+    // Check border case
+    int tmp = id + 1;
+    if (tmp == population)
+        tmp = 0;
+
+    if (lock) {
+        ownership[id] = 1;
+        ownership[tmp] = -1;
+    } else {
+        ownership[id] = 0;
+        ownership[tmp] = 0;
+    }
+
+    std::stringstream buff;
+    const int ascii_value = 65;
+
+#ifdef LOG
+    buff << "[" << id << "," << lock << "]: ";
+#endif
+
+    for (int i = 0; i < population; i++) {
+        buff << print_fork(ownership[i]);
+        buff << static_cast<char>(i + ascii_value);
+#ifdef LOG
+        buff << check_neighbour_stavation(i) << "." << starvation[i];
+#endif
+    }
+    buff << print_fork(ownership[0]);
+    buff << '\n';
+
+    std::cout << buff.str();
+}
+
+bool Table::check_neighbour_stavation(int id) {
+    int neighbour_left = id - 1;
+    int neighbour_right = id + 1;
+
+    // Check border cases
+    if(neighbour_left < 0)
+        neighbour_left = population - 1;
+    if(neighbour_right == population)
+        neighbour_right = 0;
+
+    if(starvation[neighbour_left] - starvation[id] > starvation_treshold)
+        return false;
+    if(starvation[neighbour_right] - starvation[id] > starvation_treshold)
+        return false;
+
+    return true;
+}
+
+void eat(int id, Table &table, Fork &fork_left, Fork &fork_right, std::mt19937 &seed) {
+    int result = 0;
+
+    while(result != -1) {
+        if(table.check_neighbour_stavation(id)) {
+            result = std::try_lock(fork_left.mutex, fork_right.mutex);
+        }
+
+        table.starvation[id]++;
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+
+    std::lock_guard<std::mutex> lock_left(fork_left.mutex, std::adopt_lock);
+    std::lock_guard<std::mutex> lock_right(fork_right.mutex, std::adopt_lock);
+
+    table.print(true, id);
+    table.starvation[id] = 0;
+#ifdef LOG
+    log_print(id, " started eating.");
+#endif
+    std::uniform_int_distribution<> sleep_time{1, 10};
+    std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time(seed) * 200));
+
+    table.print(false, id);
+#ifdef LOG
+    log_print(id, " finished eating.");
+#endif
+}
+
+void think(int id, std::mt19937 &seed) {
+    std::uniform_int_distribution<> sleep_time{1, 10};
+    std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time(seed) * 100));
+
+#ifdef LOG
+    log_print(id, " is done thinking ");
+#endif
```

Comment:
żeby nie robić takich ifdefów w kodzie to proponuję inne rozwiązanie. Split header / split backend.
Split header polega na tym, że log_print dajesz do pliku nagłówkowego. Są 2 wersje plików nagłówkowych - z pustą implementacją lub z wypisywaniem. Includujesz odpowiedni w zależności czy chcesz mieć printy.
Split backend jest taki, że definicja funkcji jest w nagłówku, ale są 2 różne implementacje w plikach cpp. Wtedy w CMake'u możesz jakąś flagą zdecydować który z nich ma się kompilować razem z Twoim kodem. Możesz sobie 2 targety utworzyć, jeden z printami, drugi bez.

Reply:
zostawilem te logi bo bylem ciekaw jaki jest twoj sposob na przetrzymywanie takiego kodu, ale opcja z wymiennym headerem chyba mnie troche przerasta na ten moment.